tiatoolbox/tools/patchextraction.py

```diff
@@ -0,0 +1,268 @@
+            resolution=resolution,
+            units=units,
+        )
+        self.stride = stride
+
+    def __next__(self):
+        raise NotImplementedError
+
+    def merge_patches(self, patches):
+        raise NotImplementedError
+
+
+class VariableWindowPatchExtractor(PatchExtractor):
+    """Extract and merge patches using variable sized windows for images and labels.
+
+    Args:
+        stride(Tuple of int): stride in (x, y) direction for patch extraction.
+        label_patch_size(Tuple of int): network output label (width, height).
+
+    Attributes:
+        stride(Tuple of int): stride in (x, y) direction for patch extraction.
+        label_patch_size(Tuple of int): network output label (width, height).
+
+    """
+
+    def __init__(
+        self,
+        input_img,
+        patch_size,
+        resolution=0,
+        units="level",
+        stride=1,
+        label_patch_size=None,
+    ):
+        super().__init__(
+            input_img=input_img,
+            patch_size=patch_size,
+            resolution=resolution,
+            units=units,
+        )
+        self.stride_h = stride
+        self.label_patch_size = label_patch_size
+
+    def __next__(self):
+        raise NotImplementedError
+
+    def merge_patches(self, patches):
+        raise NotImplementedError
+
+
+class PointsPatchExtractor(PatchExtractor):
+    """Extracting patches with specified points as a centre.
+
+    Args:
+        locations_list(ndarray, pd.DataFrame, str, pathlib.Path): contains location
+         and/or type of patch. Input can be path to a csv or json file.
+        num_examples_per_patch(int): Number of examples per patch for ensemble
+         classification, default=9 (centre of patch and all the eight neighbours as
+         centre).
+
+    Attributes:
+        locations_list(pd.DataFrame): A table containing location and/or type of patch.
+        num_examples_per_patch(int): Number of examples per patch for ensemble
+         classification.
+
+    """
+
+    def __init__(
+        self,
+        input_img,
+        locations_list,
+        patch_size=224,
+        resolution=0,
+        units="level",
+        num_examples_per_patch=9,
+    ):
+        super().__init__(
+            input_img=input_img,
+            patch_size=patch_size,
+            resolution=resolution,
+            units=units,
+        )
+
+        self.num_examples_per_patch = num_examples_per_patch
+        self.locations_list = read_point_annotations(input_table=locations_list)
+
+    def __next__(self):
+        n = self.n
+
+        if n >= self.locations_list.shape[0]:
+            raise StopIteration
+        self.n = n + 1
+        return self[n]
+
+    def __getitem__(self, item):
+        if type(item) is not int:
+            raise TypeError("Index should be an integer.")
+
+        if item >= self.locations_list.shape[0]:
+            raise IndexError
+
+        x, y, _ = self.locations_list.values[item, :]
```

Comment:
when adjusted `read_point_annotations` to always return class as well (None if class is absent), this line can be written like this, to support point class in the output as well (if applicable):
`x, y, c = self.locations_list.values[item, :]`

Reply:
This may generate an error due to unused variable. When we use `c` in future then we can add it. We are not using it currently therefore there is no need to assign a value to it.

Reply:
Yes you're right.
Actually, I wrote a code that extracts c for all scenarios, but I didn't push it because at my first push I realized that there are many changes required to be done in the test section to cover that and I was not sure if even you want that to be added or not!